store/daystore/day_store.go

```diff
@@ -0,0 +1,224 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package store
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"sync"
+	"time"
+
+	statstore "k8s.io/heapster/store/statstore"
+
+	"k8s.io/heapster/third_party/window"
+)
+
+// DayStore is an in-memory window of derived stats for each hour of a day.
+// DayStore holds 24 hours of derived stats, plus an hour-long StatStore that holds -
+// historical data on the current hour.
+// DayStore can calculate the Average, Max and 95th Percentile over the past 24 hours.
+// The latest one hour is only considered for the calculation of the daily Max.
```

Comment:
We can still capture spikes through the Max, which considers the current hour, and their effect on the overall daily percentile/average will be reflected when that specific hour is flushed, as of now.

If we hold 23 hourly stats and use the StatStore for the past hour then we can only guarantee that the derived stats correspond to the range of (-24,-23) hours ago. The reason is that if the last flush has happened 1 minute ago, then the StatStore holds one hour of information, 59 minutes of which we do not need. Therefore, we would calculate the average,max and 95th percentile of the past 23 hours and 1 minute, in that case. This would lead to the daily Max not reflecting a spike that has occured 23 hours and 10 minutes ago.

If we hold 24 points to avoid that and we still utilize values from the StatStore, then these values are over the 24 hour range that is stored. As we cannot account for partial data in the earliest hour, we would essentially be calculating the max, average and 95th for the past 24 hours and 1 minute instead.

Because of these reasons, I believe that the current implementation is a fair compromise of precision over time length. We use exactly 24 hours of data for the average and 95th, even though these values are at most 59 minutes in the past. For the max, we consider up to 24 hours and 59 minutes in the past, which leads to increased flexibility for spike detection. WDYT